subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java

```diff
@@ -52,7 +54,13 @@ protected CodeQualityExtension createExtension() {
         extension = project.getExtensions().create("checkstyle", CheckstyleExtension.class, project);
         extension.setToolVersion(DEFAULT_CHECKSTYLE_VERSION);
 
-        extension.setConfigDir(project.file("config/checkstyle"));
+        if (usesSubprojectCheckstyleConfiguration()) {
+            DeprecationLogger.nagUserOfDeprecated("Setting the Checkstyle configuration file under 'config/checkstyle' of a sub project", "Use the root project's 'config/checkstyle' directory instead");
+            extension.setConfigDir(project.file(CONFIG_DIR_NAME));
```

Comment:
This could be done using convention mapping instead of a setter call. Then it would be lazy, avoiding IO at configuration time.

Reply:
Done. Changed the type in the extension to `Property` and exposed a new method that takes a `Provider` so that the value be be set from the plugin implementation.